core/server/api/shared/headers.js

```diff
@@ -37,15 +37,33 @@ const disposition = {
             'Content-Type': 'application/yaml',
             'Content-Length': Buffer.byteLength(JSON.stringify(result))
         };
+    },
+
+    /**
+     * ### Content Disposition Header
+     * create a header that invokes the 'Save As' dialog in the browser when exporting the database to file. The 'filename'
+     * parameter is governed by [RFC6266](http://tools.ietf.org/html/rfc6266#section-4.3).
+     *
+     * For encoding whitespace and non-ISO-8859-1 characters, you MUST use the "filename*=" attribute, NOT "filename=".
+     * Ideally, both. Examples: http://tools.ietf.org/html/rfc6266#section-5
+     *
+     * We'll use ISO-8859-1 characters here to keep it simple.
+     *
+     * @see http://tools.ietf.org/html/rfc598
+     */
+    file(result, options, frame = {}) {
+        return {
+            'Content-Disposition': `Attachment; filename="${frame.file}"`
```

Comment:
Had tried multiple approaches and this one seems like the most appropriate in this situation. The main reason frame is used for filename storage is that we don't want to modify the api and include filename in `result` object. The second reason is that the operation of filename calculation is asynchronous and it would need a lot of reshuffling to do so (see `Object.assign(headers, disposition[apiConfig.disposition.type](result, apiConfig.disposition, frame));`). The second reason also eliminates calculating the filename directly here like csv handler does above - `if (typeof options.value === 'function') {...`

Reply:
Thanks for sharing 👍 Okay i see the problem. I am just concerned that we now use two different approaches to get header values 🤔 

Let's go back. Why is the exporter filename async? Because it reads the title from the database.
But the settings is always cached in progress. (see settings cache)

I see two other solutions:
1. the exporter uses `settingsCache.get('title')`, which makes the call sync and keeps consistenty with the current header value concept. Could be a short term solution, but not a full future proof solution.

2. As you said, make the header generation async. As soon as we have another async case we will run into this problem again. So this is a future proof solution.

---

I see two problems with attaching a value to the frame:

1. It's very hard to read. We just attach a random value `frame.file`. What does this mean? This would need a proper grouping e.g. `frame.header.value`.

2. If we want to treat header generation as a completely separate "pipe" later on, then we tie controller logic and header logic together. (headers === http, controller logic === functional/http)

---

I am fine with keeping as is + raise an issue to refactor. But i have the feeling we are raising more and more refactoring issues and we don't have much time working on them 😆

Open for any other ideas/solutions whatever 😎

Reply:
I'm trying the approach of making header assignment async. The filename generation will be similar to what `csv` header does - a call to async `options.file()` that will be defined on the controller. If that becomes too complicated will try doing the same approach but with sync `settingsCache.get('title')` call.

Reply:
The other idea would be having some extra property defined on a `frame` that would serve as a solution to keep data that dosn't really belong to the response.. but from experience those places usually become kind of a 'dump' of all sorts of things, so would rather avoid it :laughing: